File: genetic.pl
package Chromosome;

sub new {
   my $class = shift;
   my $self = {
      _string => shift,
      _score  => shift,
   };

   bless $self, $class;
   return $self;
}
sub setString {
   my ( $self, $string ) = @_;
   $self->{_string} = $string if defined($string);
   return $self->{_string};
}

sub getString {
   my( $self ) = @_;
   return $self->{_string};
}

sub setScore {
    my ( $self, $score ) = @_;
    $self->{_score} = $score if defined($score);
    return $self->{_score};
}

sub getScore {
   my( $self ) = @_;
   return $self->{_score};
}

package main;

my $low = 32;
my $high = 126;
my $POP_SIZE = 100;
my @chromosomes = ();
my @newChromosomes = ();
my @helloWorld = (72, 101, 108, 108, 111, 44, 32, 87, 111, 114, 108, 100, 33);
my $startHighestScore = 0;
my $startHighestString = "";

sub random {
    # Get number of arguments passed
    $n = scalar(@_);
    if($n == 1){
        return int(rand($_[0]));
    }else{
        return int($_[0] + rand($_[1] - $_[0]));
    }
}

sub fillArray {   
    for( my $i = 0; $i < $POP_SIZE; $i = $i + 1 ) {
        my $string = createString();
        $chromosome = new Chromosome($string, calculateFitness($string));
        push @chromosomes, $chromosome;
    }
}

sub createString {
    @string = ();
    for($i = 0; $i < 13; $i++){
        $char = chr(random($low, $high + 1));
        push @string, $char;
    }
    $str = join("", @string);
    return $str;
}

sub calculateFitness {
    my $chromeString = $_[0];
    my $score = 0;
    my @characters = split //, $chromeString;
    for($i = 0; $i < 13; $i = $i + 1){
       my $difference = abs(ord($characters[$i]) - $helloWorld[$i]);
       $score = $score + $difference;
    }
    return $score;
}

sub checkSolution {
    for($i = 0; $i < $POP_SIZE; $i++) {
        $chromosome = $chromosomes[$i];
        $chromeScore = $chromosome->getScore();
        if($chromeScore == 0){
            return 1;
        }
    }
    return 0;
}

sub setStartState {
    for($i = 0; $i < $POP_SIZE; $i++) {
        my $chromosome = $chromosomes[$i];
        my $score = $chromosome->getScore();
        my $chromeString = $chromosome->getString();
        if ($i == 0 || $score < $startHighestScore) {
            $startHighestScore = $score;
            $startHighestString = $chromeString;

        }
    }
}

sub crossOver {
    my $child1 = "";
    my $child2 = "";
    @parent1 = split //, $_[0];
    @parent2 = split //, $_[1];
    my $n = scalar(@parent1);
    my $rand = random($n);
    for ($i = 0; $i < $n; $i++) {
        if($i >= $rand) {
            $child1 .= $parent1[$i];
            $child2 .= $parent2[$i];
        } else {
            $child1 .= $parent2[$i];
            $child2 .= $parent1[$i];
        }
    }
    $ch1 = new Chromosome($child1, calculateFitness($child1));
    $ch2 = new Chromosome($child2, calculateFitness($child2));
    push @newChromosomes, $ch1;
    push @newChromosomes, $ch2;
}

sub fitnessFunction {
    $fitnessScore = 0;
    $winnerPosition = 0;
    @chromes = @_;
    for ($i = 0; $i < scalar(@chromes); $i++) {
        my $chromosome = $chromes[$i];
        my $chromeScore = $chromosome->getScore();
        if (($i == 0) || $chromeScore < $fitnessScore) {
            $winnerPosition = $i;
            $fitnessScore = $chromeScore;
        }
    }
    return $_[$winnerPosition];
}

sub createNewPopulation {
    @randomFour1 = ();
    @randomFour2 = ();

    for (my $i = 0; $i < 4; $i++) {
        my $rand1 = random($POP_SIZE);
        push @randomFour1, $chromosomes[$rand1];
        my $rand2 = random($POP_SIZE);
        push @randomFour2, $chromosomes[$rand2]; 
    }
    my $parent1Chrome = fitnessFunction(@randomFour1);
    my $parent2Chrome = fitnessFunction(@randomFour2);
    my $parent1 = $parent1Chrome->getString();
    my $parent2 = $parent2Chrome->getString();
    if (random(1, 100) < 70){
        crossOver($parent1, $parent2);
    } else {
        $ch1 = new Chromosome($parent1, calculateFitness($parent1));
        $ch1 = new Chromosome($parent2, calculateFitness($parent2));
        push @newChromosomes, $ch1;
        push @newChromosomes, $ch2;
    }
}

sub mutate {
    for (my $i = 0; $i < $POP_SIZE; $i++) {
        if (random(1, 100)< 5) {
            my $chromosome = $chromosomes[$j];
            my $stringToMutate = $chromosome->getString();
            my $fiftyFifty = random(2);
            my @characters = split //, $stringToMutate;
            my $answer = random(scalar @characters);
            my $value = ord($characters[$answer]);
            if ($value == $low) {
                $value = $value + 1;
            } elsif ($value == $high) {
                $value = $value - 1;
            } else {
                if ($fiftyFifty == 0) {
                    $value = $value + 1;
                } else {
                    $value = $value - 1;
                }
            }
            $characters[$answer] = chr($value);
            my $str = join("", @characters);
            $chromosome->setString($str);
            $chromosome->setScore(calculateFitness($str));
        }
    }
}

sub main {
    my $iteration = 0;
    fillArray();
    setStartState();
   
    while (checkSolution() == 0){
        for(my $i = 0; $i < $POP_SIZE/2; $i++){
            createNewPopulation();
            #print "$i \n";
        }
        $currentHighest = 0;
        $highestString = "";
        for (my $j = 0; $j < $POP_SIZE/2; $j++) {
            my $chromosome = $chromosomes[$j];
            my $score = $chromosome->getScore();
            my $chromeString = $chromosome->getString();
            if ($j == 0 || $score < $currentHighest) {
                $currentHighest = $score;
                $highestString = $chromeString;
            }
        }
        if (checkSolution() == 0) {
            @chromosomes = ();
            @chromosomes = @newChromosomes;
            mutate();
            print "$iteration. Best = $highestString, $currentHighest off target \n";
            @newChromosomes = ();
            $iteration++;
        }
    }
    print "Hello, World! has been made \n";
    print "Start highest string was $startHighestString ($startHighestScore away from target) \n"
}

main();

File: genetic.cpp
#include <iostream>
#include <string>
#include <random>
#include <cstdlib>
#include <ctime>
//using namespace std;

#define POP_SIZE 100
#define LOW 32
#define HIGH 126

const int helloWorld[] = {72, 101, 108, 108, 111, 44, 32, 87, 111, 114, 108, 100, 33};
const int helloWorldSize = sizeof(helloWorld)/sizeof(*helloWorld);
char randomString[helloWorldSize];
int someValues[helloWorldSize];

int startHighest = 0;
char startHighestString[helloWorldSize];

class Chromosome {
    public:
        int score;
        char string [13];

    void setScore(int score);
    void setString(char string[]);
};

Chromosome chromosomes[POP_SIZE];
Chromosome newChromosomes[POP_SIZE];

int chromosomeSize = 0;
void Chromosome::setScore( int score ) {
   this->score = score;
}
void Chromosome::setString( char chString [] ) {
    for(unsigned int i = 0; i < sizeof(*chromosomes); i++){
        string[i] = chString[i];
    }
}

int getRandom(int low, int high) {
    std::random_device _rand;
    std::mt19937 randomValue(_rand());
    std::uniform_int_distribution<> createRandom(low, high);
    return createRandom(randomValue);
}

void createString() {
    std::fill(randomString, randomString+helloWorldSize, 0);
    for(int i = 0; i < helloWorldSize; i++) {
        char charToAdd = getRandom(LOW, HIGH);
        randomString[i] = charToAdd;
    }
}

int calculateFitness(Chromosome chrome) {
    int score = 0;
    for(int i = 0; i < helloWorldSize; i++) {
        int stringValue = chrome.string[i];
        int difference = std::abs(stringValue - helloWorld[i]);
        score += difference;
    }
    return score;
}

void setStartState() {
    for(int i = 0; i < POP_SIZE; i++){
        int score = chromosomes[i].score;
        if((i == 0) || (score < startHighest)) {
            startHighest = score;
            for(int j = 0; j < helloWorldSize; j++){
                startHighestString[j] = chromosomes[i].string[j];
            }
        }
    }
}

void fillArray() {
    for(int i = 0; i < POP_SIZE; i++) {
        createString();
        Chromosome chr;
        chr.setString(randomString);
        chr.setScore(calculateFitness(chr));
        chromosomes[i] = chr;
    }
}

bool checkSolution() {
    for(int i = 0; i < POP_SIZE; i++) {
        if(chromosomes[i].score == 0) {
            return true;
        }
    }
    return false;
}

Chromosome fitnessFunction(Chromosome randomFour [], int size) {
    int fitnessScore = 0;
    int winnerPosition = 0;
    for (int i = 0; i < size; i++) { 
        if ((i == 0) || (randomFour[i].score < fitnessScore)) {
            winnerPosition = i;
            fitnessScore = randomFour[i].score;
        }
    }
    return randomFour[winnerPosition];
}

void crossOver ( char parent1 [helloWorldSize], char parent2 [helloWorldSize]) {
    int rand = getRandom(0, helloWorldSize);
    char child1 [helloWorldSize];
    char child2 [helloWorldSize];

    for (int i = 0; i < helloWorldSize; i++) {
        if (i >= rand) {
            child1[i] = parent1[i];
            child2[i] = parent2[i];
        } else {
            child1[i] = parent2[i];
            child2[i] = parent1[i];
        }
    }

    Chromosome ch1;
    Chromosome ch2;
    ch1.setString(child1);
    ch1.setScore(calculateFitness(ch1));
    newChromosomes[chromosomeSize] = ch1;
    chromosomeSize++;
    ch2.setString(child2);
    ch2.setScore(calculateFitness(ch2));
    newChromosomes[chromosomeSize] = ch2;
    chromosomeSize++;

}

void createNewPopulation() {
    Chromosome randomFour1 [4];
    Chromosome randomFour2 [4];
    int randomValue = 0;

    for(int i = 0; i < 4; i++) {
        randomValue = getRandom(0, POP_SIZE-1);
        randomFour1[i] = chromosomes[randomValue];
    }
    for(int i = 0; i < 4; i++) {
        randomValue = getRandom(0, POP_SIZE-1);
        randomFour2[i] = chromosomes[randomValue];
    }

    Chromosome parent1Chrome = fitnessFunction(randomFour1, 4);
    Chromosome parent2Chrome = fitnessFunction(randomFour2, 4);
    char parent1String [helloWorldSize];
    char parent2String [helloWorldSize];
    for (int i = 0; i < helloWorldSize; i++) {
        parent1String[i] = parent1Chrome.string[i];
        parent2String[i] = parent2Chrome.string[i];
    }
    int n = getRandom(1, 100);
    if(n < 70) {
        crossOver(parent1String, parent2String);
    } else {
        Chromosome ch1;
        Chromosome ch2;
        ch1.setString(parent1String);
        ch1.setScore(calculateFitness(ch1));
        newChromosomes[chromosomeSize] = ch1;
        chromosomeSize++;
        ch2.setString(parent2String);
        ch2.setScore(calculateFitness(ch2));
        newChromosomes[chromosomeSize] = ch2;
        chromosomeSize++;

    }
}

void mutate() {
    for (int i = 0; i < POP_SIZE; i++) {
        int n = getRandom(1, 100);
        if (n < 5){
            char stringToMuatate[helloWorldSize];
            for (int j = 0; j < helloWorldSize; j++){
                stringToMuatate[j] = chromosomes[i].string[j];
            }
            int answer = getRandom(0, helloWorldSize-1);
            int fiftyFifty = getRandom(0, 1);
            char chromosomeString = stringToMuatate[answer];
            int changedCharacter = 0;
            int value = chromosomeString;
            if (value == HIGH) {
                value++;
            } else if (value == LOW) {
                value--;
            }
            else {
                if (fiftyFifty == 0) {
                    value++;
                } else {
                    value--;
                }
            }
            char newValue = value;
            chromosomes[i].string[answer] = newValue;
            chromosomes[i].score = calculateFitness(chromosomes[i]);
        }
    }
}

int main() {
    int iteration = 0;
    fillArray();
    setStartState();
    while (!checkSolution()) {
        for(int i = 0; i < POP_SIZE/2; i++){
            createNewPopulation();
        }
        int currentHighest = 0;
        char highestString[helloWorldSize + 1];
        for(int i = 0; i < POP_SIZE; i++){
            int score = chromosomes[i].score;
            if ((i == 0) || (score < currentHighest)) {
                currentHighest = score;
                for (int j = 0; j < helloWorldSize; j++) {
                    highestString[j] = chromosomes[i].string[j];
                }
            } 
            highestString[helloWorldSize] = '\0';  

        }
        if (!checkSolution()) {
            for(int i = 0; i < POP_SIZE; i++){
                chromosomes[i] = newChromosomes[i];
            }
            mutate();
            chromosomeSize = 0;
            iteration ++;
            std::cout << iteration << ". Best = " << highestString << " (" << currentHighest << " away from target)" << std::endl;
        }
        
    }
    std::cout << "Hello, World! has been made" << "\n\n";
    std::cout << "Start string was " << startHighestString << " (" << startHighest <<" away from target)" << "\n\n";
    return 0;
}
